js-bmfbp/prototype.js

```diff
@@ -0,0 +1,273 @@
+//
+// The Parts
+
+// Part A: It keeps an incrementing count starting from 0 every second,
+// multiplied by 2 and send it to its OUT pin.
+function partA(part, send, releaseDeferred) {
+  var count = 0;
+
+  setInterval(function () {
+    count++;
+    // TODO: Future suggestion: Allow sending to a pin name instead of a pin
+    // index.
+    send(part, 0, count * 2);
+    releaseDeferred(part);
+  }, 1000);
+
+  // This two-layer definition is necessary because everything in JS is wrapped
+  // in a lambda and we don't have access to mutable variables that survive
+  // across function calls.
+  return function (pin, packet) {
+  };
+}
+
+// Part B: It keeps an incrementing count starting from 0 every second,
+// multiplied by 3 and send it to its OUT pin.
+function partB(part, send, releaseDeferred) {
+  var count = 0;
+
+  setInterval(function () {
+    count++;
+    send(part, 0, count * 3);
+    releaseDeferred(part);
+  }, 1000);
+
+  return function (pin, packet) {
+  };
+}
+
+// Part C: It receives a count from its IN pin and creates an object with the
+// key of "count" and the value as the count from the IN pin. It then sends the
+// object to its first OUT pin. If the count is odd, send another packet with a
+// message to its second OUT pin.
+function partC(part, send) {
+  return function (pin, packet) {
+    switch (pin) {
+      case 0:
+        send(part, 0, {
+          "count": packet
+        });
+
+        if (packet % 2 != 0) {
+          send(part, 1, {
+            "message": "Odd number detected",
+            "count": packet
+          });
+        }
+    }
+  };
+}
+
+// Part D: It prints packets from its IN pin to the console.
+function partD(part, send) {
+  return function (pin, packet) {
+    switch (pin) {
+      case 0:
+        console.log(packet);
+    }
+  };
+}
+
+// Part E: It takes the message in the packet from its IN pin, adds one to the
+// count property, then print it to the console, prefixed with "ADDED ONE: "
+function partE(part, send) {
+  return function (pin, packet) {
+    switch (pin) {
+      case 0:
+        // TODO: Currently this would increment the count for `partF` as well
+        // because there is no copying. This is an issue to be addressed.
+        packet.count++;
+        console.log("ADDED ONE: " + packet.count);
+    }
+  };
+}
+
+// Part F: It prints the message in the packet from its IN pin to the console,
+// prefixed with "WARNING: ".
+function partF(part, send) {
+  return function (pin, packet) {
+    switch (pin) {
+      case 0:
+        console.log("WARNING: " + packet.message + ": " + packet.count);
+    }
+  };
+}
+
+// ----------------------------------------------------------------------------
+//
+// The schematic, conceptually:
+//
+//     B --0--+
+//            |
+//            v
+//     A --1--+-> C --2---> D
+//                |
+//                +-+--3--> E
+//                  +--4--> F
```

Comment:
https://github.com/bmfbp/bmfbp/pull/11/commits/64bca9a3e3f37973566a07d5757e6eec3dbb6a67

Reply:
@guitarvydas Actually, I get that conceptually having one wire here vs distinguishing this into two wires are the same. Is there any reason against distinguishing it into two wires? It makes the code easier to reason about.

Reply:
In electronics, it is one wire - the signal arrives at E & F "at the same time" (when "time" is much greater than the speed of light ; there did exist a class of circuits that were so fast that the Engineer needed to calculate propagation delays and compensate for wire lengths).  And so, I find it easier to reason about as one wire.

In software, the "at the same time" consideration still holds.  3 & 4 imply copying.  During copying, 3 & 4 must "lock" the input queues of E & F so that no other event can wriggle through and change the order of events.

Let's say that Part C shoots two events, X & Y, in rapid succession.  We must guarantee that, both, E & F "see" event X before seeing event Y (X is in E & F's input queues as #1, Y is #2 in both queues - always).

In electronics, the propagation delay depends on the length of the wire (I think it's 9 inches for one nanosecond).  In software, the propagation delay is the time it takes to lock all input queues, do the copying and push the events onto the queue.

Hence 3 & 4 act "differently" (slower) than 1 & 2.  As long as we remember that semantic restriction, it doesn't matter how we label the line segments (I would say that the labels are 3a, 3b and 3c, but I don't really care what the labels are).

Reply:
I *think* that the "at the same time" semantic is how we get rid of a certain class of race conditions.